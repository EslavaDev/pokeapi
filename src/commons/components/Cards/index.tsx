import {useQuery} from '@tanstack/react-query';
import React, {useCallback} from 'react';
import {View, Text, Image, StyleSheet, TouchableOpacity} from 'react-native';
import {useNavigation} from '@react-navigation/native';

export const Card = ({pokemon, queryKey}: any) => {
  console.log("ðŸš€ ~ Card ~ pokemon:", pokemon)
  const navigation = useNavigation();
  const {data} = useQuery({
    queryKey: [queryKey, pokemon.name],
    queryFn: () => fetch(pokemon.url).then(res => res.json()),
    enabled: !!pokemon?.sprites,
  });

  const pokemonData = data || pokemon;

  const onPress = useCallback(() => {
    navigation.navigate('Detail', {pokemon: pokemonData});
  }, [navigation, pokemonData]);
  return (
    <TouchableOpacity onPress={onPress} style={styles.card}>
      <Image
        source={{uri: pokemonData?.sprites?.front_default}}
        style={styles.image}
      />
      <Text style={styles.name}>{pokemonData?.name}</Text>
      <View style={styles.info}>
        <Text style={styles.text}>#{pokemonData?.order}</Text>
        {/* <Text style={styles.text}>Type: {data?.types?.map(item => item.type.name).join(", ")}</Text>
        <Text style={styles.text}>HP: {data?.stats[0]?.base_stat}</Text> */}
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  card: {
    flex:1,
    maxHeight: 250,
    backgroundColor: '#f0f0f0',
    borderRadius: 10,
    overflow: 'hidden',
    borderColor: '#ddd',
    borderWidth: 1,
    margin: 5,
    elevation: 3,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
  },
  image: {
    width: '100%',
    height: 150,
    backgroundColor: 'red',
    resizeMode: 'contain',
  },
  name: {
    fontWeight: 'bold',
    fontSize: 18,
    marginTop: 0,
    textAlign: 'center',
    color: '#333',
  },
  info: {
    padding: 10,
  },
  text: {
    fontSize: 14,
    color: '#555',
    textAlign: 'center',
  },
});
