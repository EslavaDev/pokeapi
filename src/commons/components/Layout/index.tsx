import React from 'react';
import {FlexStyle, SafeAreaView, ScrollView} from 'react-native';
import {KeyboardAwareScrollView} from 'react-native-keyboard-aware-scroll-view';
import {stylesLayout} from './styles';

interface Props {
  children: React.ReactNode;
  keyboard?: boolean;
  position?: FlexStyle['position'];
  justifyContent?: FlexStyle['justifyContent'];
}

export const Layout = ({
  children,
  justifyContent = 'center',
  position = 'relative',
  keyboard = false,
}: Props) => {
  const styles = stylesLayout({justifyContent, position});
  let render = (
    <ScrollView>
      <SafeAreaView style={styles.container}>{children}</SafeAreaView>
    </ScrollView>
  );
  if (keyboard) {
    render = (
      <KeyboardAwareScrollView
        style={styles.keyboard}
        contentContainerStyle={styles.container}>
        <SafeAreaView style={[styles.container]}>{children}</SafeAreaView>
      </KeyboardAwareScrollView>
    );
  }
  return render;
};
