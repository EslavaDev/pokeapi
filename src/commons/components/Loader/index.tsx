import { ActivityIndicator, StyleSheet, View } from "react-native";
type Props = {
    validation: boolean;
}
export const Loader = ({validation}: Props) => {
    if(!validation) return null;
    return (
        <View style={styles.container}>
        <ActivityIndicator size="large" color="#0000ff" />
        </View>
    );
    }

const styles = StyleSheet.create({
    container: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    });
