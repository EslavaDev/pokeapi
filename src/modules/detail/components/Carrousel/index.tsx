import React from 'react';
import {FlatList, Image, View} from 'react-native';
import styles from './styles';

type Props = {
  pokemon: any;
};
export const Carrousel = ({pokemon}: Props) => {
  const _renderItem = ({item}: any) => {
    return (
      <View>
        <Image
          source={{
            uri: item,
          }}
          style={styles.image}
        />
      </View>
    );
  };

  const imagePokemon = Object.values(pokemon?.sprites || {}).filter(
    (item: any) => typeof item === 'string' && item.includes('https'),
  );

  return (
    <View style={styles.container}>
      <FlatList data={imagePokemon} renderItem={_renderItem} horizontal />
    </View>
  );
};
