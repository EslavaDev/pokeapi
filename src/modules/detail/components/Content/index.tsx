import React, {PropsWithChildren} from 'react';
import {Dimensions, StyleSheet, View} from 'react-native';

type Props = PropsWithChildren<{}>;

export const ContentLayout = ({children}: Props) => (
  <View style={styles.layout}>{children}</View>
);

const styles = StyleSheet.create({
  layout: {
    flex: 1,
    width: Dimensions.get('window').width,
    flexDirection: 'column',
    backgroundColor: 'green',
  },
});
