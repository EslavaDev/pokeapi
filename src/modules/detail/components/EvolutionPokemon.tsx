import {pokemonFullDataService} from '@src/modules/home/services';
import {useQuery} from '@tanstack/react-query';
import {View} from 'react-native';
import {Text} from 'react-native-svg';
import {evolveRecursion} from '../utils';

export const EvolutionPokemon = ({url}: any) => {
  const {data: evolution} = useQuery({
    queryKey: ['evolution', url],
    queryFn: pokemonFullDataService(url),
    select: data =>
      data.chain.evolves_to.map((result: any) => [
        data.chain.species.name,
        ...evolveRecursion(result),
      ]),
  });

  return (
    <View style={{flex: 1, backgroundColor: 'yellow'}}>
      <Text style={{fontSize: 20}}>{evolution?.join(' => ')}</Text>
    </View>
  );
};
