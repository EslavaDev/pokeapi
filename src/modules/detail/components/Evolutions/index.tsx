import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import {EvolutionByPokemon} from './EvolutionByPokemon';

type Props = {
  evolution: any;
};

export const Evolutions = ({evolution}: Props) => (
  <View style={styles.container}>
    <Text style={styles.text}>Evoluciones</Text>
    <EvolutionByPokemon url={evolution?.evolution_chain?.url} />
  </View>
);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f0f0f0',
    borderRadius: 10,
    borderColor: '#7a7a7a',
    borderWidth: 1,
    margin: 5,
    padding: 10,
    elevation: 3,
    gap: 15,
    shadowColor: '#000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    width: '50%',
  },
  text: {
    fontWeight: 'bold',
    fontSize: 25,
    textAlign: 'center',
    textTransform: 'capitalize',
  },
});
