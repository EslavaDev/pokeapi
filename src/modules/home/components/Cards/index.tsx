import {useQuery} from '@tanstack/react-query';
import React, {useCallback} from 'react';
import {View, Text, Image, TouchableOpacity} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import styles from './styles';
import {pokemonFullDataService} from '../../../../services';

export const Card = ({pokemon, queryKey}: any) => {
  const navigation = useNavigation<any>();
  const {data} = useQuery({
    queryKey: [queryKey, pokemon.name],
    queryFn: pokemonFullDataService(pokemon?.url ?? ''),
    enabled: !pokemon?.sprites,
  });

  const pokemonData = data || pokemon;

  const onPress = useCallback(() => {
    navigation.navigate('Detail', {pokemon: pokemonData});
  }, [navigation, pokemonData]);

  return (
    <TouchableOpacity onPress={onPress} style={styles.card}>
      <Image
        source={{uri: pokemonData?.sprites?.front_default}}
        style={styles.image}
      />
      <Text style={styles.name}>{pokemonData?.name}</Text>
      <View style={styles.info}>
        <Text style={styles.text}>#{pokemonData?.id}</Text>
      </View>
    </TouchableOpacity>
  );
};
