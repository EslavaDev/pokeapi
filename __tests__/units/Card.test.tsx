/**
 * @format
 */

import 'react-native';
import React from 'react';

// Note: import explicitly to use the types shipped with jest.
import {it, expect} from '@jest/globals';

import renderer from 'react-test-renderer';
import {useQuery} from '@tanstack/react-query';
import {useNavigation} from '@react-navigation/native';
import {Card} from '@src/modules/home/components/Cards';
import {TouchableOpacity} from 'react-native';

jest.mock('@tanstack/react-query', () => ({
  useQuery: jest.fn(() => ({
    data: {
      name: 'Pikachu',
      sprites: {
        front_default: 'https://pokeapi.co/media/sprites/pokemon/25.png',
      },
      id: 25,
    },
  })),
}));
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(() => ({
    navigate: jest.fn(),
  })),
}));

it('should navigate to the Detail screen when the card is pressed', () => {
  const pokemon = {
    name: 'Pikachu',
    sprites: {
      front_default: 'https://pokeapi.co/media/sprites/pokemon/25.png',
    },
    id: 25,
  };

  const mockNavigation = {
    navigate: jest.fn(),
  };
  useNavigation.mockReturnValue(mockNavigation);

  const component = renderer.create(
    <Card pokemon={pokemon} queryKey="pokemon" />,
  );

  component.root.findByType(TouchableOpacity).props.onPress();

  expect(mockNavigation.navigate).toHaveBeenCalledWith('Detail', {
    pokemon: pokemon,
  });
});

it('should render the card component without a pokemon object', () => {
  const component = renderer.create(<Card queryKey="pokemon" />);

  expect(component.toJSON()).toMatchSnapshot();
});
