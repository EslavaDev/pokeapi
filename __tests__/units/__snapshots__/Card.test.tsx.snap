// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the card component without a pokemon object 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#f0f0f0",
      "borderColor": "#7a7a7a",
      "borderRadius": 10,
      "borderWidth": 1,
      "elevation": 3,
      "flex": 1,
      "margin": 5,
      "maxHeight": 210,
      "opacity": 1,
      "overflow": "hidden",
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.25,
      "shadowRadius": 3.84,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://pokeapi.co/media/sprites/pokemon/25.png",
      }
    }
    style={
      {
        "backgroundColor": "#b9b8ae",
        "height": 150,
        "resizeMode": "contain",
        "width": "100%",
      }
    }
  />
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 18,
        "fontWeight": "bold",
        "marginTop": 3,
        "textAlign": "center",
      }
    }
  >
    Pikachu
  </Text>
  <View
    style={
      {
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "#555",
          "fontSize": 14,
          "textAlign": "center",
        }
      }
    >
      #
      25
    </Text>
  </View>
</View>
`;
